runtime/lua/vim/lsp/completion.lua

```diff
@@ -462,7 +461,9 @@ local function trigger(bufnr, clients)
       end
     end
     local start_col = (server_start_boundary or word_boundary) + 1
-    vim.fn.complete(start_col, matches)
+    vim.schedule(function()
+      vim.fn.complete(start_col, matches)
```

Comment:
is there a 1-line comment that can give a hint about why the schedule() is needed?

Reply:
It raises a `Vim:E565: Not allowed to change text or change window`  without it.

I suspect this doesn't happen outside of test-cases because the rpc layer already adds a schedule to escape the luv callback.

We could probably also do something like this:

```diff
diff --git a/test/functional/plugin/lsp/completion_spec.lua b/test/functional/plugin/lsp/completion_spec.lua
index 7e2dd0fbe3..a25bc738c9 100644
--- a/test/functional/plugin/lsp/completion_spec.lua
+++ b/test/functional/plugin/lsp/completion_spec.lua
@@ -729,13 +729,10 @@ end)
 
 describe('vim.lsp.completion: integration', function()
   before_each(function()
     clear()
     exec_lua(create_server_definition)
-    exec_lua(function()
-      vim.fn.complete = vim.schedule_wrap(vim.fn.complete)
-    end)
   end)
 
   after_each(clear)
 
   it('puts cursor at the end of completed word', function()
diff --git a/test/functional/plugin/lsp/testutil.lua b/test/functional/plugin/lsp/testutil.lua
index a36cbac568..ca00805a16 100644
--- a/test/functional/plugin/lsp/testutil.lua
+++ b/test/functional/plugin/lsp/testutil.lua
@@ -62,10 +62,11 @@ M.create_server_definition = function()
       function srv.request(method, params, callback)
         table.insert(server.messages, {
           method = method,
           params = params,
         })
+        callback = vim.schedule_wrap(callback)
         local handler = handlers[method]
         if handler then
           handler(method, params, callback)
         elseif method == 'initialize' then
           callback(nil, {
```

Reply:
I'd tend to, alone for the reason to bring the in-process test server closer to the behavior of a cmd-stdio/tcp server.

And for future people + test-cases - debugging that this was necessary took quite some time as it's non-obvious and the errror wasn't propagated.

Reply:
Looks like making it general would break a couple of other tests due to the changed timing.
Given the impact I'd rather change this separately so I'm going ahead and merge this as is.